Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp

```diff
@@ -0,0 +1,283 @@
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebExtensionRegisteredScriptsSQLiteStore.h"
+
+#include "Logging.h"
+#include "WebExtensionSQLiteHelpers.h"
+#include <wtf/JSONValues.h>
+#include <wtf/TZoneMallocInlines.h>
+#include <wtf/text/MakeString.h>
+
+namespace WebKit {
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+WTF_MAKE_TZONE_ALLOCATED_IMPL(WebExtensionRegisteredScriptsSQLiteStore);
+
+static const SchemaVersion currentSchemaVersion = 1;
+
+static constexpr auto idKey = "id"_s;
+static constexpr auto persistAcrossSessionsKey = "persistAcrossSessions"_s;
+
+static String rowFilterStringFromRowKeys(Vector<String> keys)
+{
+    Vector<String> escapedAndQuotedKeys;
+    for (String key : keys) {
+        String keyWithSingleQuotesEscaped = makeStringByReplacingAll(key, "'"_s, "''"_s);
+        escapedAndQuotedKeys.append(makeString("'"_s, keyWithSingleQuotesEscaped, "'"_s));
+    }
+    return makeStringByJoining(escapedAndQuotedKeys.span(), ", "_s);
+}
+
+WebExtensionRegisteredScriptsSQLiteStore::WebExtensionRegisteredScriptsSQLiteStore(const String& uniqueIdentifier, const String& directory, bool useInMemoryDatabase)
+    : WebExtensionSQLiteStore(uniqueIdentifier, directory, useInMemoryDatabase)
+{
+}
+
+void WebExtensionRegisteredScriptsSQLiteStore::updateScripts(Vector<Ref<JSON::Object>> scripts, CompletionHandler<void(const String& errorMessage)>&& completionHandler)
+{
+    Vector<String> ids;
+    for (const auto& script : scripts)
+        ids.append(script->getString(idKey));
+
+    deleteScriptsWithIDs(ids, [weakThis = ThreadSafeWeakPtr { *this }, ids = WTFMove(ids), scripts = WTFMove(scripts), completionHandler = WTFMove(completionHandler)](const String& errorMessage) mutable {
+        RefPtr protectedThis = weakThis.get();
+        if (!protectedThis) {
+            completionHandler({ });
+            return;
+        }
+
+        if (!errorMessage.isEmpty()) {
+            completionHandler(errorMessage);
+            return;
+        }
+
+        protectedThis->addScripts(scripts, [completionHandler = WTFMove(completionHandler)](const String& errorMessage) mutable {
+            completionHandler(errorMessage);
+        });
+    });
+}
+
+void WebExtensionRegisteredScriptsSQLiteStore::deleteScriptsWithIDs(Vector<String> ids, CompletionHandler<void(const String& errorMessage)>&& completionHandler)
+{
+    if (ids.isEmpty()) {
+        completionHandler({ });
+        return;
+    }
+
+    queue()->dispatch([weakThis = ThreadSafeWeakPtr { *this }, ids = WTFMove(ids), completionHandler = WTFMove(completionHandler)]() mutable {
```

Comment:
All the queue dispatches need to use `crossThreadCopy` instead of `WTFMove`. See https://commits.webkit.org/290441@main.
```suggestion
    queue()->dispatch([weakThis = ThreadSafeWeakPtr { *this }, ids = crossThreadCopy(ids), completionHandler = WTFMove(completionHandler)]() mutable {
```